perfkitbenchmarker/windows_virtual_machine.py

```diff
@@ -0,0 +1,306 @@
+
+    delete_connection = 'net use %s /delete' % root
+
+    cmd = ';'.join([set_error_pref, create_cred, create_psdrive,
+                    copy_item, delete_connection])
+
+    stdout, stderr, retcode = vm_util.IssueCommand(
+        ['powershell', '-Command', cmd], timeout=None)
+
+    if retcode:
+      error_text = ('Got non-zero return code (%s) executing %s\n'
+                    'STDOUT: %sSTDERR: %s' %
+                    (retcode, cmd, stdout, stderr))
+      raise errors.VirtualMachine.RemoteCommandError(error_text)
+
+  @vm_util.Retry(log_errors=False, poll_interval=1)
+  def WaitForBootCompletion(self):
+    """Waits until VM is has booted."""
+    resp, _ = self.RemoteCommand('hostname', suppress_warning=True)
+    if self.bootable_time is None:
+      self.bootable_time = time.time()
+    if self.hostname is None:
+      self.hostname = resp[:-1]
+
+  def OnStartup(self):
+    stdout, _ = self.RemoteCommand('echo $env:TEMP')
+    self.temp_dir = os.path.join(stdout.strip(), 'pkb')
+    self.RemoteCommand('mkdir %s' % self.temp_dir)
+    self.DisableGuestFirewall()
+
+  def Install(self, package_name):
+    """Installs a PerfKit package on the VM."""
+    if ((self.is_static and not self.install_packages) or
+        not FLAGS.install_packages):
+      return
+    if package_name not in self._installed_packages:
+      package = windows_packages.PACKAGES[package_name]
+      package.Install(self)
+      self._installed_packages.add(package_name)
+
+  def Uninstall(self, package_name):
+    """Uninstalls a Perfkit package on the VM."""
+    package = windows_packages.PACKAGES[package_name]
+    if hasattr(package, 'Uninstall'):
+      package.Uninstall()
+
+  def PackageCleanup(self):
+    """Cleans up all installed packages.
+
+    Deletes the temp directory and uninstalls all PerfKit packages.
+    """
+    for package_name in self._installed_packages:
+      self.Uninstall(package_name)
+    self.RemoteCommand('rm -recurse -force %s' % self.temp_dir)
+    self.DisableGuestFirewall()
+
+  def _GetNumCpus(self):
+    """Returns the number of logical CPUs on the VM.
+
+    This method does not cache results (unlike "num_cpus").
+    """
+    stdout, _ = self.RemoteCommand(
+        'Get-WmiObject -class Win32_processor | '
+        'select -exp NumberOfLogicalProcessors')
+    return int(stdout)
+
+  def _GetTotalMemoryKb(self):
+    """Returns the amount of physical memory on the VM in Kilobytes.
+
+    This method does not cache results (unlike "total_memory_kb").
+    """
+    stdout, _ = self.RemoteCommand(
+        'Get-WmiObject -class Win32_PhysicalMemory | '
+        'select -exp Capacity')
+    return int(stdout) / 1024
+
+  def _TestReachable(self, ip):
+    """Returns True if the VM can reach the ip address and False otherwise."""
+    try:
+      self.RemoteCommand('ping -n 1 %s' % ip)
+    except errors.VirtualMachine.RemoteCommandError:
+      return False
```

Comment:
If the first VM couldn't be reached we wouldn't have reached this point in the benchmark. Unless something really wrong happens (like the VM goes down), the only reason we'd fail here is if the internal ip isn't reachable. If a VM goes down, I'd argue we have larger problems than thinking we can't reach the internal ip address.